### ISSUE TITLE
Add a config option to disable proxying for all private packages (to shield against dependency confusion attacks)

### ISSUE THREAD

**Issue by Cin316** [verdaccio/verdaccio#2103]
**Is your feature request related to a problem?**
- [Dependency confusion](https://medium.com/@alex.birsan/dependency-confusion-4a5d60fec610) attacks exploit a vulnerability in private package registry configurations where an attacker creates a public package with the same name as a private package.
- I believe Verdaccio installations may be at risk to these attacks, but I don't understand the particulars of Verdaccio's package resolution algorithm.
- What package will Verdaccio serve in this situation?
  - I have a private package called "my-private-package" (version 1.0.0) in my Verdaccio registry.
  - There is a public package with the same name at version 2.0.0
- Dependency confusion attacks work when the private registry serves the public package instead of the private one.

**Describe the solution you'd like**
- I would like a config option, which when enabled, prevents requests for any packages which exist in a private Verdaccio installation from going out to public NPM.

**Describe alternatives you've considered**
- I understand that you can set the "packages" config so that certain package name patterns don't proxy out to the public NPM.  However, this requires that all your private package names follow a consistent pattern, which may not always be the case.

**Additional context**
- I can work on this feature if you give me some direction.

**Comment by DanielRuf** [comment]
See https://verdaccio.org/docs/en/best#remove-proxy-to-increase-security-at-private-packages

**Comment by DanielRuf** [comment]
>     * I understand that you can set the "packages" config so that certain package name patterns don't proxy out to the public NPM.  However, this requires that all your private package names follow a consistent pattern, which may not always be the case.

In general this is a best practice (scopes).

**Comment by Cin316** [comment]
@DanielRuf I saw that, but is there a way to automatically make sure all private packages do not proxy, but all other package names do proxy?  It looks like you have to write matching patterns for all of your private package names in the config to only disable the proxy for them.

**Comment by juanpicado** [comment]
> but is there a way to automatically make sure all private packages do not proxy, 

no yet, but I was thinking the other day we might check which packages are in the `.verdaccio-db.json` file and avoid proxy any private package by default.

**Comment by Cin316** [comment]
@juanpicado I think that would be a great addition and would solve the problem I described above.

**Comment by juanpicado** [comment]
> @juanpicado I think that would be a great addition and would solve the problem I described above.

Unique corner case I did not give many thoughts is the compatibility with all storage plugins, maybe you might check the code, only two files do the job (local-storage, and storage) and share the outcome afterward. I can provide you guidence.

**Comment by juanpicado** [comment]
One more corner case, people that have already published eg: (webpack patch) the package will be present at `.verdaccio-db.json`, thus won't be proxied anymore. So, in any case, this feature should be under a flag until a new release major is available.

**Comment by amitgilad3** [comment]
I dived into the issue and reached a few interesting conclusions:


Possible solution:

1. Implement flag that will block all published packages from being resolved by uplink.

problem:  when proxy is enabled for all then every time a package is published verdaccio will verify this against the uplink ,and if the version exists it will fail.

possible solutions :
1.  if you publish higher version, then from that point on verdaccio will stop resolving for that package(which will cause all sort of issues stated by @juanpicado  ).
> One more corner case, people that have already published eg: (webpack patch) the package will be present at `.verdaccio-db.json`, thus won't be proxied anymore.

2. disable proxy for all - but then you don't need this flag - just create a private registry for private packages and a public one for the rest

3.  i think that the best solution is not implementing anything and using @DanielRuf solution which seems to be the best solution i see
> See https://verdaccio.org/docs/en/best#remove-proxy-to-increase-security-at-private-packages

problem: 
handling backward compatibility for organization that already have verdaccio with private packages ?

if anyone has a better solution please let me know and i will try to implement it otherwise i think using https://verdaccio.org/docs/en/best#remove-proxy-to-increase-security-at-private-packages is the best and we can close the issue

**Comment by juanpicado** [comment]
It is a though decision, since the nature of the proxy is just expect that works. I'd go with number 3. But the original request is legitimate at least to consider a flag or something to disable all proxies by default.

Time ago I was thinking on add `-offline` flag https://github.com/verdaccio/verdaccio/discussions/1742 and that give an answer to this request. Thoughts?

**Comment by amitgilad3** [comment]
> It is a though decision, since the nature of the proxy is just expect that works. I'd go with number 3. But the original request is legitimate at least to consider a flag or something to disable all proxies by default.
> 
> Time ago I was thinking on add `-offline` flag #1742 and that give an answer to this request. Thoughts?

@juanpicado 
so just to clarify, this flag would disable all proxying  and ensure that only packages in the registry are served ?? 
and if so  i would be glad to start working on it if nobody started ??

**Comment by juanpicado** [comment]
>  this flag would disable all proxying and ensure that only packages in the registry are served ??

Yes, in theory you only need to block two paths, `getTarball()` and `getPackage()` in the storage. 

I haven't purposed any implementation here https://github.com/verdaccio/verdaccio/discussions/1742 just a flag. But we need to link the flag to the configuration file, in other words, like `--listen` also can be set up on the `config.yaml` file, this can be done on `@verdaccio/cli` module and set the prop to the parsed config file, furthermore involves update `@verdaccio/types`, and maybe `@verdaccio/config`.

As you might now, we have few properties as `url_prefix`, `max_body_size`, `http_proxy` or `no_proxy` are set on the root. I don't want to do this anymore to avoid future collisions with plugins #928 . I'd ask to use `server` namespaces for the `offline` mode or if you have a different idea we can discuss it.

Another thing, I'd go only with `master` branch, I don't want to have more features on `5.x` since is more work in maintenance and I'd like to release alpha late this year. Furthermore, it's much easier to create test at the current modules, 5.x is a bit of a mess tbh.

Let's try to create this web framework agnostic as possible 🙏🏼 , since I'm working on parallel to migrate to fastify.

**Comment by juanpicado** [comment]
One little improvement is document better in the config file about this topic (that would included in the next release  v5.10.0
https://github.com/verdaccio/verdaccio/commit/5c650acbb204b71cbf1da65148ca570bba8484fd#diff-d3b601db7d2bbb294e8d7a367f2aedbd8c19c65e5c62ba3ea49c38d9f72862caR99

**Comment by mbtools** [comment]
Using Verdaccio as an [offline repository](https://verdaccio.org/docs/linking-remote-registry#offline-registry) is simply a config with no or with an empty `uplinks` section:

```yml
uplinks:
```

Alternatively, remove all `proxy` values from the `packages` section:

```yml
packages:
  '@*/*':
    access: $all
    publish: $authenticated
  '**':
    access: $all
    publish: $authenticated
```

There's also an [option](https://github.com/verdaccio/verdaccio/pull/4687) in v8.x to disable the uplinks tab on the web ui:

```yml
web: 
  showUplinks: false
```

The [publish option](https://verdaccio.org/docs/next/configuration/#offline-publish) can also be relevant: 

```yml
pulish: 
  allow_offline: true
```

These settings will make sure that all packages come from Verdaccio only and prevent any dependency confusion attacks.
